src/main/java/org/jenkinsci/plugins/junitrealtimetestreporter/AbstractRealtimeTestResultAction.java

```diff
@@ -0,0 +1,121 @@
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+package org.jenkinsci.plugins.junitrealtimetestreporter;
+
+import hudson.AbortException;
+import hudson.Main;
+import hudson.model.Run;
+import hudson.tasks.test.AbstractTestResultAction;
+import hudson.tasks.test.TestResult;
+import java.io.IOException;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+import org.kohsuke.stapler.HttpRedirect;
+import org.kohsuke.stapler.StaplerProxy;
+
+abstract class AbstractRealtimeTestResultAction extends AbstractTestResultAction<AbstractRealtimeTestResultAction> implements StaplerProxy {
+
+    private static final Logger LOGGER = Logger.getLogger(AbstractRealtimeTestResultAction.class.getName());
+
+    protected TestResult result;
+    private transient long updated;
+
+    protected AbstractRealtimeTestResultAction() {}
+
+    protected abstract TestResult parse() throws IOException, InterruptedException;
+
+    @Override
+    public TestResult getResult() {
+        // Refresh every 1/100 of a job estimated duration but not more often than every 5 seconds
+        final long threshold = Math.max(5000, run.getEstimatedDuration() / 100);
+        if (updated > System.currentTimeMillis() - threshold && !Main.isUnitTest) {
+            LOGGER.fine("Cache hit");
+            return result;
+        }
+        try {
+            final long started = System.currentTimeMillis(); // TODO use nanoTime
+            // TODO this can block on Remoting and hang the UI; need to refresh results asynchronously
+            result = parse();
+            result.setParentAction(this);
+            LOGGER.log(Level.FINE, "Parsing took {0} ms", System.currentTimeMillis() - started);
+            updated = System.currentTimeMillis();
+        } catch (AbortException ex) {
+            // Thrown when there are no reports or no workspace witch is normal
+            // at the beginning the build. This is also a signal that there are
+            // no reports to update (already parsed was excluded and no new have
+            // arrived so far).
+            LOGGER.fine("No new reports found.");
```

Comment:
Sounds a bit hackish. Would it be possible to propagate the cases above as a special exception type? Otherwise a real issue may be suppressed in a custom `parse()` implementation. 🐜

Reply:
Again, perhaps, but this was code which was already there and I am touching unnecessarily.